flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/controller/FlinkDeploymentController.java

```diff
@@ -113,29 +118,70 @@ public DeleteControl cleanup(FlinkDeployment flinkApp, Context context) {
         Configuration effectiveConfig =
                 FlinkUtils.getEffectiveConfig(flinkApp, defaultConfig.getFlinkConfig());
 
-        boolean readyToReconcile = observer.observe(flinkApp, context, effectiveConfig);
-        if (!readyToReconcile) {
-            return flinkApp.getStatus()
-                    .getJobManagerDeploymentStatus()
-                    .toUpdateControl(flinkApp, operatorConfiguration);
-        }
-
         try {
+            boolean readyToReconcile = observer.observe(flinkApp, context, effectiveConfig);
+            if (!readyToReconcile) {
+                return flinkApp.getStatus()
+                        .getJobManagerDeploymentStatus()
+                        .toUpdateControl(flinkApp, operatorConfiguration);
+            }
+
             UpdateControl<FlinkDeployment> updateControl =
                     reconcilerFactory
                             .getOrCreate(flinkApp)
                             .reconcile(operatorNamespace, flinkApp, context, effectiveConfig);
             updateForReconciliationSuccess(flinkApp);
             return updateControl;
-        } catch (InvalidDeploymentException ide) {
-            LOG.error("Reconciliation failed", ide);
-            updateForReconciliationError(flinkApp, ide.getMessage());
+        } catch (DeploymentFailedException dfe) {
+            updateForDeploymentFailed(flinkApp, dfe);
             return UpdateControl.updateStatus(flinkApp);
         } catch (Exception e) {
             throw new ReconciliationException(e);
         }
     }
 
+    private void updateForDeploymentFailed(
+            FlinkDeployment flinkApp, DeploymentFailedException dfe) {
+        LOG.error(
+                "Deployment {}/{} failed with {}",
+                flinkApp.getMetadata().getNamespace(),
+                flinkApp.getMetadata().getName(),
+                dfe.getMessage());
+        flinkApp.getStatus().setJobManagerDeploymentStatus(JobManagerDeploymentStatus.ERROR);
+        updateForReconciliationError(flinkApp, dfe.getMessage());
+
+        // TODO: avoid repeated event
+        EventBuilder evtb =
+                new EventBuilder()
+                        .withApiVersion("v1")
```

Comment:
Coud this use the ApiVersion of the FlinkDeployment?

Reply:
No, that version is `flink.apache.org/v1alpha1`